k8s-operator/apis/v1alpha1/types_proxygroup.go

```diff
@@ -49,27 +54,60 @@ type ProxyGroupSpec struct {
 	// Replicas specifies how many replicas to create the StatefulSet with.
 	// Defaults to 2.
 	// +optional
-	Replicas int `json:"replicas,omitempty"`
+	Replicas *int `json:"replicas,omitempty"`
+
+	// HostnamePrefix is the hostname prefix to use for tailnet devices created
+	// by the ProxyGroup. Each device will have the integer number from its
+	// StatefulSet pod appended to this prefix to form the full hostname.
+	// HostnamePrefix can contain lower case letters, numbers and dashes, it
+	// must not start with a dash and must be between 1 and 62 characters long.
+	// +optional
+	HostnamePrefix HostnamePrefix `json:"hostnamePrefix,omitempty"`
 
-	// ProxyClass is the name of the ProxyClass custom resource that
-	// contains configuration options that should be applied to the
-	// resources created for this ProxyGroup. If unset, the operator will
-	// create resources with the default configuration.
+	// ProxyClass is the name of the ProxyClass custom resource that contains
+	// configuration options that should be applied to the resources created
+	// for this ProxyGroup. If unset, and no default ProxyClass is set, the
+	// operator will create resources with the default configuration.
 	// +optional
 	ProxyClass string `json:"proxyClass,omitempty"`
 }
 
 type ProxyGroupStatus struct {
-	// List of status conditions to indicate the status of the ProxyGroup pods.
-	// Known condition types are `ProxyGroupReady`.
+	// List of status conditions to indicate the status of the ProxyGroup
+	// resources. Known condition types are `ProxyGroupReady`.
 	// +listType=map
 	// +listMapKey=type
 	// +optional
 	Conditions []metav1.Condition `json:"conditions,omitempty"`
 
-	// List of tailnet devices associated with the ProxyGroup statefulset.
+	// List of tailnet devices associated with the ProxyGroup StatefulSet.
 	// +listType=map
 	// +listMapKey=hostname
 	// +optional
 	Devices []TailnetDevice `json:"devices,omitempty"`
 }
+
+type TailnetDevice struct {
+	// Hostname is the fully qualified domain name of the device.
+	// If MagicDNS is enabled in your tailnet, it is the MagicDNS name of the
+	// node.
+	Hostname string `json:"hostname"`
+
+	// TailnetIPs is the set of tailnet IP addresses (both IPv4 and IPv6)
+	// assigned to the device.
+	// +optional
+	TailnetIPs []string `json:"tailnetIPs,omitempty"`
+}
+
+// +kubebuilder:validation:Type=string
+// +kubebuilder:validation:Pattern=`^[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$`
```

Comment:
I think this should be an [enum](https://book.kubebuilder.io/reference/markers/crd-validation#:~:text=values%20specified%20here.-,//%20%2Bkubebuilder%3Avalidation%3AEnum,this%20(scalar)%20field%20is%20restricted%20to%20the%20*exact*%20values%20specified%20here.,-//%20%2Bkubebuilder%3Avalidation%3AExclusiveMaximum), we will only allow one of a specific set of values
Should be  `// +kubebuilder:validation:Enum=egress` (I assume we don't want to default to egress)

Reply:
Oops not sure how I managed that, could have sworn I had written it as an enum already. Thanks for catching.